lib/unitary_calculator_basic.h

```diff
@@ -197,209 +161,76 @@ class UnitaryCalculatorBasic final {
-          dot_two_r(m02_r, m02_i, m12_r, m12_i, m22_r, m22_i, m32_r,
-                    m32_i, 2, matrix);
-      data[2 * p * sizei + 2 * pp + 1] =
-          dot_two_i(m02_r, m02_i, m12_r, m12_i, m22_r, m22_i, m32_r,
-                    m32_i, 2, matrix);
-      pp |= sizek;
-      data[2 * p * sizei + 2 * pp] =
-          dot_two_r(m03_r, m03_i, m13_r, m13_i, m23_r, m23_i, m33_r,
-                    m33_i, 2, matrix);
-      data[2 * p * sizei + 2 * pp + 1] =
-          dot_two_i(m03_r, m03_i, m13_r, m13_i, m23_r, m23_i, m33_r,
-                    m33_i, 2, matrix);
-
-      p |= sizek;
-      pp = si2;
-      data[2 * p * sizei + 2 * pp] =
-          dot_two_r(m00_r, m00_i, m10_r, m10_i, m20_r, m20_i, m30_r,
-                    m30_i, 3, matrix);
-      data[2 * p * sizei + 2 * pp + 1] =
-          dot_two_i(m00_r, m00_i, m10_r, m10_i, m20_r, m20_i, m30_r,
-                    m30_i, 3, matrix);
-      pp = si2 | sizek;
-      data[2 * p * sizei + 2 * pp] =
-          dot_two_r(m01_r, m01_i, m11_r, m11_i, m21_r, m21_i, m31_r,
-                    m31_i, 3, matrix);
-      data[2 * p * sizei + 2 * pp + 1] =
-          dot_two_i(m01_r, m01_i, m11_r, m11_i, m21_r, m21_i, m31_r,
-                    m31_i, 3, matrix);
-      pp = si2 | sizej;
-      data[2 * p * sizei + 2 * pp] =
-          dot_two_r(m02_r, m02_i, m12_r, m12_i, m22_r, m22_i, m32_r,
-                    m32_i, 3, matrix);
-      data[2 * p * sizei + 2 * pp + 1] =
-          dot_two_i(m02_r, m02_i, m12_r, m12_i, m22_r, m22_i, m32_r,
-                    m32_i, 3, matrix);
-      pp |= sizek;
-      data[2 * p * sizei + 2 * pp] =
-          dot_two_r(m03_r, m03_i, m13_r, m13_i, m23_r, m23_i, m33_r,
-                    m33_i, 3, matrix);
-      data[2 * p * sizei + 2 * pp + 1] =
-          dot_two_i(m03_r, m03_i, m13_r, m13_i, m23_r, m23_i, m33_r,
-                    m33_i, 3, matrix);
+    uint64_t xs[2];
+    uint64_t ms[3];
+
+    xs[0] = uint64_t{1} << (q0 + 1);
+    ms[0] = (uint64_t{1} << q0) - 1;
+    for (unsigned i = 1; i < 2; ++i) {
+      xs[i] = uint64_t{1} << (q1 + 1);
+      ms[i] = ((uint64_t{1} << q1) - 1) ^ (xs[i - 1] - 1);
+    }
+    ms[2] = ((uint64_t{1} << num_qubits_) - 1) ^ (xs[1] - 1);
+
+    uint64_t xss[4];
+    for (unsigned i = 0; i < 4; ++i) {
+      uint64_t a = 0;
+      for (uint64_t k = 0; k < 2; ++k) {
+        if (((i >> k) & 1) == 1) {
+          a += xs[k];
+        }
+      }
+      xss[i] = a;
+    }
+
+    auto f = [q0, q1](unsigned n, unsigned m, uint64_t ii, const fp_type* v,
+                      const uint64_t* ms, const uint64_t* xss, unsigned n_qb,
+                      unsigned sqrt_size, fp_type* rstate) {
+      fp_type rn, in;
+      fp_type rs[16], is[16];
+
+      auto row_size = uint64_t{1} << n_qb;
+      uint64_t i = ii % sqrt_size;
+      uint64_t j = ii / sqrt_size;
+
+      uint64_t k = (1 * i & ms[0]) | (2 * i & ms[1]) | (4 * i & ms[2]);
+      uint64_t kk = (1 * j & ms[0]) | (2 * j & ms[1]) | (4 * j & ms[2]);
```

Comment:
Same here - rename to avoid collision with `k` used in loop.

Reply:
Done.